### PR TITLE
feat: multi tenancy basic configuration

### PR DESCRIPTION
**Pull request by wojcikmat**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?

It creates a new app for multi tenancy purposes. It's a basic functionalities (kind of draft) of this feature. Introduces Tenant model, connection to User and add a simple behavior to create default user's tenant during sign up.

### What is the current behavior?

Currently there is no multi tenancy feature

### Does this PR introduce a breaking change?

Not for now.

### Other information:

As for now it is first PR connected with multi tenancy topic. There will be much more, but for the purpose of reviewing smaller parts of code and making sure that the direction is fine there will be smaller PRs to the main feature branch, which will be eventually merged to master branch.

### CODE REVIEW

File: packages/backend/apps/multitenancy/constants.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from django.db import models
+
+
+class TenantType(models.TextChoices):
+    SIGN_UP = "sign_up", "Sign Up"
```
Comment (pziemkowski) [inline packages/backend/apps/multitenancy/constants.py:5]: Do you have a better idea how to name this? 
I'm kind of afraid using a verb as a noun is going to be confusing for all people.

I was confused even though I'm pretty sure I suggested the name :D 

(Can't be `default` either IMO)

Comment (wojcikmat) [reply packages/backend/apps/multitenancy/constants.py:5]: sure, tbh I was thinking about `default` but your suggestion make me think that maybe `sign_up` will be better name :D

File: packages/backend/apps/multitenancy/models.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+from .managers import TenantManager
+
+
+class Tenant(models.Model):
+    id: str = hashid_field.HashidAutoField(primary_key=True)
+    creator: settings.AUTH_USER_MODEL = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    name: str = models.CharField(max_length=100, unique=True)
+    slug: str = models.SlugField(max_length=100, unique=True)
+    type: str = models.CharField(choices=constants.TenantType.choices)
+    created = models.DateTimeField(auto_now_add=True)
```
Comment (pziemkowski) [inline packages/backend/apps/multitenancy/models.py:17]: I think it should be `created_at`?

and `updated_at` while you're at it. I wonder if we don't have a mixin for this already. If not, pretty sure it's a good idea to introduce it.

File: packages/backend/apps/multitenancy/models.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        return self.name
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = slugify(self.name)
+        super().save(*args, **kwargs)
+
+
+class TenantMembership(models.Model):
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name="memberships")
```
Comment (pziemkowski) [inline packages/backend/apps/multitenancy/models.py:34]: what do you think about more descriptive `tenant_memberships` related name for user model?

File: packages/backend/apps/multitenancy/tests/factories.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import factory
```
Comment (pziemkowski) [inline packages/backend/apps/multitenancy/tests/factories.py:1]: I might be getting too far ahead with this but maybe we should automatically create a tenant + membership when you create a user with UserFactory?

**Review by pziemkowski — CHANGES_REQUESTED** [review]

File: packages/backend/apps/multitenancy/constants.py (hunk 1 above)
Comment (pziemkowski) [reply packages/backend/apps/multitenancy/constants.py:5]: But `default` has its own problems. Could be interpreted as `default` selection when logging in.

Comment (pziemkowski) [reply packages/backend/apps/multitenancy/constants.py:5]: Anyway, I checked and Netlify calls it `default` as well

**Review by pziemkowski — APPROVED** [review]

File: packages/backend/apps/multitenancy/models.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+from . import constants
+from .managers import TenantManager
+from common.models import TimestampedMixin
+
+
+class Tenant(TimestampedMixin, models.Model):
+    id: str = hashid_field.HashidAutoField(primary_key=True)
+    creator: settings.AUTH_USER_MODEL = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
+    name: str = models.CharField(max_length=100, unique=True)
+    slug: str = models.SlugField(max_length=100, unique=True)
```
Comment (mkleszcz) [inline packages/backend/apps/multitenancy/models.py:16]: I think we need to add a unique index to this field

File: packages/backend/apps/multitenancy/models.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    )
+
+    objects = TenantManager()
+
+    def __str__(self):
+        return self.name
+
+    def save(self, *args, **kwargs):
+        if not self.slug:
+            self.slug = slugify(self.name)
```
Comment (mkleszcz) [inline packages/backend/apps/multitenancy/models.py:29]: This needs to check if the slug is already in the db and iterate to find unique one

**Review by mkleszcz — CHANGES_REQUESTED** [review]

**Review by mkleszcz — APPROVED** [review]

**Pull request merged**